### PR TITLE
Add modules needed for LUKS

### PR DESCRIPTION
**Pull request by doozan**
This adds the modules needed to use LUKS.

When `CONFIG_ARM64_CRYPTO` is defined, the macro `module_cpu_feature_match()` is used to check that the CPU supports the crypto extension.  Unfortunately, that macro won't compile and gives the error:

```
In file included from arch/arm64/crypto/aes-glue.c:16:0:
include/linux/module.h:87:32: error: storage size of '__mod_cpu_device_table' isn't known
 extern const struct gtype##_id __mod_##gtype##_table  \
                                ^
include/linux/module.h:145:3: note: in expansion of macro 'MODULE_GENERIC_TABLE'
   MODULE_GENERIC_TABLE(type##_device, name)
   ^~~~~~~~~~~~~~~~~~~~
include/linux/cpufeature.h:49:1: note: in expansion of macro 'MODULE_DEVICE_TABLE'
 MODULE_DEVICE_TABLE(cpu, cpu_feature_match_ ## x);  \
 ^~~~~~~~~~~~~~~~~~~
arch/arm64/crypto/aes-glue.c:442:1: note: in expansion of macro 'module_cpu_feature_match'
 module_cpu_feature_match(AES, aes_init);
 ^~~~~~~~~~~~~~~~~~~~~~~~
make[1]: *** [/src/CoreELEC-coreelec-9.0/build.CoreELEC-Amlogic.arm-9.0-devel/linux-f1bb00821950b0f5cab1b8a820c697ba9fbd3183/arch/arm64/crypto/Makefile:38: arch/arm64/crypto/aes-glue-ce.o] Error 1
```

I've included a patch to remove the macro. Since we're targeting a specific device and it will have the expected extensions it should be safe, but I'm open to suggestions on how to better fix this.

The patch is important, because without the arm crypto, it's pretty slow.

Before:

```
CoreELEC:~ # cryptsetup benchmark
# Tests are approximate using memory only (no storage IO).
PBKDF2-sha1       281875 iterations per second for 256-bit key
PBKDF2-sha256     569878 iterations per second for 256-bit key
PBKDF2-sha512     238312 iterations per second for 256-bit key
PBKDF2-ripemd160  152409 iterations per second for 256-bit key
PBKDF2-whirlpool     N/A
argon2i       4 iterations, 169514 memory, 4 parallel threads (CPUs) for 256-bit key (requested 2000 ms time)
argon2id      4 iterations, 186603 memory, 4 parallel threads (CPUs) for 256-bit key (requested 2000 ms time)
#     Algorithm |       Key |      Encryption |      Decryption
        aes-cbc        128b        37.0 MiB/s        38.7 MiB/s
    serpent-cbc        128b               N/A               N/A
    twofish-cbc        128b               N/A               N/A
        aes-cbc        256b        29.5 MiB/s        30.4 MiB/s
    serpent-cbc        256b               N/A               N/A
    twofish-cbc        256b               N/A               N/A
        aes-xts        256b        39.9 MiB/s        39.6 MiB/s
    serpent-xts        256b               N/A               N/A
    twofish-xts        256b               N/A               N/A
        aes-xts        512b        31.0 MiB/s        28.6 MiB/s
    serpent-xts        512b               N/A               N/A
    twofish-xts        512b               N/A               N/A
```

After:

```
CoreELEC:~ # modprobe aes_ce_blk
CoreELEC:~ # cryptsetup benchmark
# Tests are approximate using memory only (no storage IO).
PBKDF2-sha1       281875 iterations per second for 256-bit key
PBKDF2-sha256     582542 iterations per second for 256-bit key
PBKDF2-sha512     238312 iterations per second for 256-bit key
PBKDF2-ripemd160  152409 iterations per second for 256-bit key
PBKDF2-whirlpool     N/A
argon2i       4 iterations, 168870 memory, 4 parallel threads (CPUs) for 256-bit key (requested 2000 ms time)
argon2id      4 iterations, 173398 memory, 4 parallel threads (CPUs) for 256-bit key (requested 2000 ms time)
#     Algorithm |       Key |      Encryption |      Decryption
        aes-cbc        128b       282.5 MiB/s       284.2 MiB/s
    serpent-cbc        128b               N/A               N/A
    twofish-cbc        128b               N/A               N/A
        aes-cbc        256b       244.6 MiB/s       244.8 MiB/s
    serpent-cbc        256b               N/A               N/A
    twofish-cbc        256b               N/A               N/A
        aes-xts        256b       264.9 MiB/s       265.1 MiB/s
    serpent-xts        256b               N/A               N/A
    twofish-xts        256b               N/A               N/A
        aes-xts        512b       235.6 MiB/s       236.3 MiB/s
    serpent-xts        512b               N/A               N/A
    twofish-xts        512b               N/A               N/A
```

### CODE REVIEW

**Comment by Ray-future** [comment]
The Linux patch should go into the Linux branch. Other than that I would be OK with the change if it was up to me.

**Comment by afl1** [comment]
I think is better fix this issue, as hack code which is part of mainline linux. This issue was fix in 4.9 kernel by removing _id in MODULE_GENERIC_TABLE macro.[module: fix types of device tables aliases ](https://github.com/CoreELEC/linux-amlogic/commit/6301939d97d079f0d3dbe71e750f4daf5d39fc33#diff-4b68ce3c700ded96468c46aa417fb441)

**Comment by doozan** [comment]
Thank you for pointing me in the right direction @afl1, that's a much better fix!

I've added packages for cryptsetup and lvm2 and put cryptsetup on the busybox dependency list.  This will put everything needed to boot with an encrypted /storage partition into the default image.

The user will need to supply his own mount-storage.sh to handle unlocking, but I'll post an example here as a proof-of-concept.  This script temporarily sets up system, loads the kernel modules, and starts SSH waiting for the user to connect and unlock the storage.

To unlock the device remotely, run `ssh root@coreelec.ip.address 'echo -n MYPASSWORD | /bin/cryptsetup luksOpen /dev/MY_LUKS_DEVICE storage-crypt -d - && killall sshd'`

/flash/mount-storage.sh
```
#!/bin/sh

echo "setting up temp SYSTEM"

mount_part "/flash/$IMAGE_SYSTEM" "/sysroot" "ro,loop"
/usr/bin/busybox mount --bind /dev /sysroot/dev
/usr/bin/busybox mount --bind /proc /sysroot/proc
/usr/bin/busybox mount --bind /sys /sysroot/sys
/usr/bin/busybox mount --bind /flash /sysroot/flash
mount -t tmpfs tmpfs /sysroot/var
mount -t tmpfs tmpfs /sysroot/run
mount -t tmpfs tmpfs /sysroot/storage

mkdir /sysroot/var/lib
chroot /sysroot /usr/bin/ln -s /usr/lib/kernel-overlays/base/lib/modules /var/lib/modules
chroot /sysroot /usr/sbin/modprobe dm-crypt
chroot /sysroot /usr/sbin/modprobe aes-ce-blk
chroot /sysroot /usr/sbin/modprobe algif_skcipher


echo "setting up sshd"

mkdir -p /sysroot/var/empty
mkdir -p /sysroot/storage/.cache/ssh/

# This is unencrypted, don't re-use the hostkey from your encrypted /storage/.ssh folder
cat<<'EOF'>/sysroot/storage/.cache/ssh/ssh_host_ed25519_key
PASTE_SSH_HOST_KEY_HERE
EOF
/sysroot/usr/bin/busybox chmod 600 /sysroot/storage/.cache/ssh/ssh_host_ed25519_key


# Uncomment this if you want to allow password login (password is corelec)
#cat<<'EOF'>/sysroot/storage/.cache/shadow
#root:$6$j2kltGrL/rskio2D$Leynsm6Dk/TBBxsdhRCNqjjwy4wnfwshN6lDnrNKMGejw9qVjlnwhQlmUYekno.oSpcDpPRxJNkUCwGJrGxLj0:::::::
#EOF
#/sysroot/usr/bin/busybox chmod 600 /sysroot/storage/.cache/shadow

mkdir /sysroot/storage/.ssh
cat<<'EOF'>/sysroot/storage/.ssh/authorized_keys
PASTE_SSH_PUBLIC_KEY_HERE
EOF
/sysroot/usr/bin/busybox chmod 600 /sysroot/storage/.ssh/authorized_keys

echo "Starting ssh"
chroot /sysroot /usr/sbin/sshd -D

# User must connect via ssh, unlock storage as storage-crypt and kill sshd

echo "Cleaning up temp system"
umount /sysroot/run
umount /sysroot/flash
umount /sysroot/dev
umount /sysroot/proc
umount /sysroot/sys
umount /sysroot/var

umount /sysroot/storage
umount /sysroot

if [ -b /dev/mapper/storage-crypt ]; then
  echo "Mounting encrypted storage..."
  mount_part "/dev/mapper/storage-crypt" "/storage" "rw,noatime"
else
  echo "storage-crypt missing, mounting default storage..."
  mount_part "$disk" "/storage" "rw,noatime"
fi
```

File: packages/linux/patches/linux-010_fix_device_table_aliases.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff -ur a/include/linux/module.h b/include/linux/module.h
```
Comment (afl1) [inline packages/linux/patches/linux-010_fix_device_table_aliases.patch]: Pls, remove linux patch as I fixed it in amlogic-3.14 kernel.
[https://github.com/CoreELEC/linux-amlogic/commit/a76ebe491925b300b97bef6a88b818d1f82038a4](https://github.com/CoreELEC/linux-amlogic/commit/a76ebe491925b300b97bef6a88b818d1f82038a4)

Comment (doozan) [reply packages/linux/patches/linux-010_fix_device_table_aliases.patch]: > Pls, remove linux patch as I fixed it in amlogic-3.14 kernel.

Done!

**Comment by doozan** [comment]
I'm still interesting in seeing this merged, is there anything more needed from my side?

It seems like it should be pretty straight forward to include the first patch so that we include the same kernel modules that are already being built for the Allwinner/Rockchip configs.

**Comment by spycat88** [comment]
Apologies that this has been left for many months, I'm good with accepting this into CE but your cryptsetup and lvm2 package.mk are not in good alignment with the rest of the buildroot packages, there is a template you can look at [here](https://raw.githubusercontent.com/CoreELEC/CoreELEC/coreelec-9.2/packages/packages.mk.template).

I would also suggest submitting the PR upstream to LE for further review.

**Comment by jedi7** [comment]
Hi, is there any progress? Do you need a help?

**Comment by doozan** [comment]
I opened #230 with a patch to include the same kernel modules that are already being built for the Allwinner/Rockchip configs.  That's the minimum needed to make LUKS viable on CoreELEC.  If/when that gets merged, I'll work on up-streaming the other patches to build the user land binaries

**Pull request closed without merging**